rmw_opendds_cpp/src/rmw_service.cpp

```diff
@@ -26,319 +26,241 @@
+  if (publisher->set_default_datawriter_qos(writer_qos) != DDS::RETCODE_OK) {
+    RMW_SET_ERROR_MSG("failed to set_default_datawriter_qos");
+    return nullptr;
   }
-/* //commented out for clearing crashes before type support is ready.
-   //TODO: implement service properly.
-  replier = callbacks->create_replier(
-    participant, request_topic_str, response_topic_str,
-    &datareader_qos, &datawriter_qos,
-    reinterpret_cast<void **>(&request_datareader),
-    reinterpret_cast<void **>(&response_datawriter),
-    &rmw_allocate);
 
-  CORBA::string_free(request_topic_str);
-  request_topic_str = nullptr;
-  CORBA::string_free(response_topic_str);
-  response_topic_str = nullptr;
-
-  if (!replier) {
-    RMW_SET_ERROR_MSG("failed to create replier");
-    goto fail;
-  }
-  if (!request_datareader) {
-    RMW_SET_ERROR_MSG("data reader handle is null");
-    goto fail;
+  // create subscriber
+  DDS::Subscriber_var subscriber = node_info->participant->create_subscriber(
+    SUBSCRIBER_QOS_DEFAULT, NULL, OpenDDS::DCPS::NO_STATUS_MASK);
+  if (!subscriber) {
+    RMW_SET_ERROR_MSG("failed to create_subscriber");
+    return nullptr;
   }
-  if (!response_datawriter) {
-    RMW_SET_ERROR_MSG("data writer handle is null");
-    goto fail;
+  DDS::DataReaderQos reader_qos;
+  if (!get_datareader_qos(subscriber, *qos_profile, reader_qos)) {
+    RMW_SET_ERROR_MSG("failed to get_datareader_qos");
+    return nullptr;
   }
-
-  read_condition = request_datareader->create_readcondition(
-    DDS::ANY_SAMPLE_STATE, DDS::ANY_VIEW_STATE, DDS::ANY_INSTANCE_STATE);
-  if (!read_condition) {
-    RMW_SET_ERROR_MSG("failed to create read condition");
-    goto fail;
+  if (subscriber->set_default_datareader_qos(reader_qos) != DDS::RETCODE_OK) {
+    RMW_SET_ERROR_MSG("failed to set_default_datareader_qos");
+    return nullptr;
   }
 
-  dds_subscriber = request_datareader->get_subscriber();
-  status = participant->get_default_subscriber_qos(subscriber_qos);
-  if (status != DDS::RETCODE_OK) {
-    RMW_SET_ERROR_MSG("failed to get default subscriber qos");
-    goto fail;
+  // create service
+  rmw_service_t * service = rmw_service_allocate();
+  if (!service) {
+    RMW_SET_ERROR_MSG("failed to allocate service");
+    return nullptr;
   }
+  void * buf = nullptr;
+  try {
+    service->implementation_identifier = opendds_identifier;
+    service->data = nullptr;
+    service->service_name = reinterpret_cast<const char *>(rmw_allocate(strlen(service_name) + 1));
+    if (!service->service_name) {
+      throw std::string("failed to allocate memory for service name");
+    }
+    memcpy(const_cast<char*>(service->service_name), service_name, strlen(service_name) + 1);
 
-  dds_publisher = response_datawriter->get_publisher();
-  status = participant->get_default_publisher_qos(publisher_qos);
-  if (status != DDS::RETCODE_OK) {
-    RMW_SET_ERROR_MSG("failed to get default subscriber qos");
-    goto fail;
-  }
+    // create OpenDDSStaticServiceInfo
```

Comment:
It is pretty clear what is happening based on the next line, so I'd remove this comment.

Reply:
done